Added two particle generators

hlokavarapu opened the pull request:
PR includes two new particle generators 'reference cell' and 'quadrature points' as was suggested by @gassmoeller, and @egpuckett via email. 

'Reference cell' generates particles uniformly distributed in the unit cell and transforms them back to the real cell. 

The second particle generator 'quadrature points,' generates particles at quadrature points of each active cell in the triangulation mesh.

This PR also includes tests for both generators. 

However, the particle generators are partially incomplete because producing unique IDs for particles is difficult if they are generated an a cell by cell basis. For example, iterating over each active cell on a local MPI process, the quadrature points are initialized and particles are generated locally at these points. As they are generated, to obtain a unique id, one needs knowledge of the number of particles that have been already been generated on neighbouring MPI processes. 

I have come up with a possible solution but I am afraid it will require a bit of debugging. The basic idea is to pass a token from MPI process to MPI process that indicates the starting ID to being from given that each MPI process knows exactly the number of total particles to generate locally. Once all MPI processes know their given starting point, at which point they begin generating particles. I can continue to pursue this if others think that this is the right approach.

bangerth commented:
@hlokavarapu -- the usual approach is to let every processor generate its own particles first, enumerating them starting at zero. Then every process sends its number of particles to all other processes, which can then add up how many particles all of the processes with lower number have together. Let's say that on process $P$ this number is $N_P$ where of course $N_0=0$. Process P then adds $N_P$ to the ids of all of the particles that it owns.

You can make your life a bit easier by combining the data exchange and formation of the partial sum -- the combined operation is called a "scan". An example of a use is in `source/particle/generator/probability_density_function.cc`.

gassmoeller commented:
Nice work, those generators are going to be useful! 
As an addition to Wolfgang's comment: In order to generate the particles first and then later on change their id, you would probably need to add a `set_id` function to the `Particle` class (I am not opposed to that). But in the other generators we determine the local starting index before actually generating the particles and then generate them with their proper id. This is a bit more in line with the idea of a constant, unique id for a particle that never changes (although on construction I would not oppose changing their ids around a bit). 
You could actually also do the same for these generators, because you know in advance that you are going to generate `n_quadrature_points * this->get_triangulation().n_locally_active_cells()` and `n_particles_per_cell * this->get_triangulation().n_locally_active_cells() ` particles respectively. So you could do the `MPI_Scan` operation beforehand, and then generate all particles with their correct id, this is also how it is done in `source/particle/generator/probability_density_function.cc`

hlokavarapu commented:
Using the scan operation, support for unique particle indices has been added for both generators. Updated tests to reflect this. And lastly, added support for unique particle indices where `generate_particles()` is called more than once.

hlokavarapu commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -29,62 +29,85 @@ namespace aspect
-                 q_points_itr != quadrature_points.end(); q_points_itr++)
-              {
-                Point<dim> particle_position_real = (*q_points_itr);
+            n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * fe_values.n_quadrature_points;
```
Here, we compute the total number of particles to generate locally for each MPI process.

hlokavarapu commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -29,62 +29,85 @@ namespace aspect
+            n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * fe_values.n_quadrature_points;
+          }
+
+        MPI_Scan(&n_particles_to_generate, &prefix_sum, 1, MPI_UNSIGNED, MPI_SUM, this->get_mpi_communicator());
```
Given the number of particles to generate locally, compute the appropriate offset in the global generation of particles and store the result in variable prefix_sum.

hlokavarapu commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -29,62 +29,85 @@ namespace aspect
               }
           }
+
+        starting_particle_index = Utilities::MPI::max(particle_index, this->get_mpi_communicator());
```
Lastly, update the global starting particle index such that if this function is called again, unique particle indices can still be maintained.

gassmoeller commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -0,0 +1,130 @@
+        if (cell != endc)
+          {
+            fe_values.reinit(cell);
+            n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * fe_values.n_quadrature_points;
```
You should be able to get the number of quadrature point from the quadrature formula (`quadrature_formula.size()`). That way you would not need to check for `cell != endc` and you do not need to initialize `fe_values` here.

gassmoeller commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -0,0 +1,130 @@
+            n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * fe_values.n_quadrature_points;
+          }
+
+        MPI_Scan(&n_particles_to_generate, &prefix_sum, 1, MPI_UNSIGNED, MPI_SUM, this->get_mpi_communicator());
```
Could you change `MPI_UNSIGNED` to `ASPECT_TRACER_INDEX_MPI_TYPE`. Although no individual process will own more than 4 billion particles (although it is possible on machines with 512 GB RAM, I tested that once :smile:) the overall sum can be larger.

gassmoeller commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -0,0 +1,130 @@
+      {
+        const QGauss<dim> quadrature_formula(this->get_parameters().stokes_velocity_degree + 1);
+
+        unsigned int n_particles_to_generate = 0, prefix_sum = 0;
```
Also please change the type to `types::particle_index`, see below for a justification

gassmoeller commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -0,0 +1,130 @@
+                  {
+                    Point<dim> particle_position_real = (*q_points_itr);
+
+                    try
```
This try-catch block should be unnecessary. The quadrature points will always be in the unit cell, and `get_quadrature_points` will always return positions in the real cell, if not something is seriously wrong and we should just throw the exception. In other words `particle_position_unit` for particle number `i` of a cell is always equal to `quadrature_formula.get_points[i]` so you do not need to compute it again.

gassmoeller commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -0,0 +1,130 @@
+      ASPECT_REGISTER_PARTICLE_GENERATOR(QuadraturePoints,
+                                         "quadrature points",
+                                         "Generate particles at the quadrature points of each active"
+                                         "cell of the triangulation mesh.")
```
Same comment applies as for the class description in the header.

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+/*
+  Copyright (C) 2015 by the authors of the ASPECT code.
```
2016

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+      template <int dim>
+      ReferenceCell<dim>::ReferenceCell()
+      {
+        starting_particle_index = 0;
```
initializer-list

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+        const std::vector<Point<dim>> particles_in_unit_cell = generate_particle_positions_in_unit_cell();
+
+        unsigned int n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * particles_in_unit_cell.size();
+        unsigned int prefix_sum = 0;
```
make both types `types::particle_index`

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+        unsigned int n_particles_to_generate = this->get_triangulation().n_locally_owned_active_cells() * particles_in_unit_cell.size();
+        unsigned int prefix_sum = 0;
+
+        MPI_Scan(&n_particles_to_generate, &prefix_sum, 1, MPI_UNSIGNED, MPI_SUM, this->get_mpi_communicator());
```
`ASPECT_TRACER_MPI_TYPE`

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+                     itr_particles_in_unit_cell != particles_in_unit_cell.end();
+                     itr_particles_in_unit_cell++)
+                  {
+                    try
```
You can remove the try-catch block. As long as your `generate_particle_positions_in_unit_cell` function does its job correctly, there should not be any problem (and if it is not, there should be an assertion inside that function that the generated points lie inside the unit cell). The interior of the try block is correct.

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+                  {
+                    try
+                      {
+                        Point<dim> position_real = this->get_mapping().transform_unit_to_real_cell(cell,
```
you can make this `const`

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+      void
+      ReferenceCell<dim>::generate_particles(std::multimap<types::LevelInd, Particle<dim> > &particles)
+      {
+        const std::vector<Point<dim>> particles_in_unit_cell = generate_particle_positions_in_unit_cell();
```
please put a space between the two closing `>` `>`, older compilers misinterpret this sometimes as a `>>` operator (we had this problem a while ago).

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+
+
+      template <int dim>
+      const std::vector<Point<dim>>
```
`> >`

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+      const std::vector<Point<dim>>
+                                 ReferenceCell<dim>::generate_particle_positions_in_unit_cell()
+      {
+        std::vector<Point<dim>> particle_positions;
```
same here

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+              {
+                if (dim == 2)
+                  {
+                    Point<dim> position_unit = Point<dim>(i * spacing[0] + spacing[0] / 2,
```
`const`

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+                  {
+                    for (unsigned int k = 0; k < number_of_particles[2]; ++k)
+                      {
+                        Point<dim> position_unit = Point<dim>(i * spacing[0] + spacing[0] / 2,
```
`const`

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+            {
+              prm.enter_subsection("Reference cell");
+              {
+                prm.declare_entry ("Number of tracers per cell per direction", "10",
```
From looking at the code I was thinking one would be able to prescribe different numbers of particle per direction. I do not know if anybody would want to do this, but in case you would be interested: There is a function called `Utilities::possibly_extend_from_1_to_N` that would allow you to either enter one number (for the same number of particles in every direction), or a list of `dim` entries for the different directions. It is for example used for the `densities` parameter of the `DiffusionDislocation` material model. If you are not interested, you can also keep everything as is (it just looked like you prepared the code, and then did not make the input parameter for different numbers).

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,192 @@
+    {
+      ASPECT_REGISTER_PARTICLE_GENERATOR(ReferenceCell,
+                                         "reference cell",
+                                         "Generates a uniform distribution of particles "
```
I would add `per cell and spatial direction` to avoid the impression you have to provide `n_particles` locations

gassmoeller commented on tests/particle_generator_reference_cell.prm:
```diff
@@ -0,0 +1,122 @@
+
+    subsection Generator
+      subsection Reference cell
+        set Number of tracers per cell per direction = 0.4e1
```
Would you mind changing this to 4? :smirk:

gassmoeller commented:
/run-tests

bangerth commented on include/aspect/particle/generator/reference_cell.h:
```diff
@@ -0,0 +1,90 @@
+          void
+          generate_particles(std::multimap<types::LevelInd, Particle<dim> > &particles);
+
+          const std::vector<Point<dim>> generate_particle_positions_in_unit_cell();
```
The `const` at the beginning of the line is meaningless here, so you may as well omit it.

Also, please put a space between `>>` -- it is required for C++ versions previous to C++11.

hlokavarapu commented:
Thanks Rene. I will run several scaling tests in the near future but based on a quick test on my local machine, I did verify that as the number of MPI processes increase for a fixed number of particles, the wall clock time decreases proportionally. Really happy with these generators, wall clock time wise.

The recent commit resolves the various comments regarding code style, optimization, compiler bugs and suggested code functionality.

hlokavarapu commented:
I am a bit puzzled by the failing tests. After reproducing the failed tests using ctest on my local machine, I have noticed that the diff command executed by ctest on my machine is:

`diff /home/harsha_lv/CIG_codes/build/aspect/new_particle_generators/tests/output-particle_generator_reference_cell/particle-00000.0000.txt /home/harsha_lv/ClionProjects/aspect/tests/particle_generator_reference_cell/particle-00000.0000.txt`

The problem is that the first file does not exist. The file is not being copied over to the tests/output-particle_generator_reference_cell/ after the output has been generated. For example, /home/harsha_lv/CIG_codes/build/aspect/new_particle_generators/tests/output-particle_generator_reference_cell/particles/particle-00000.0000.txt exist. The diff between /home/harsha_lv/CIG_codes/build/aspect/new_particle_generators/tests/output-particle_generator_reference_cell/particles/particle-00000.0000.txt and that of the uploaded test output files particle-00000.0000.txt is zero.

@tjhei Do you know what might be happening here?

How does ctest know to copy the output data from <output_dir>/particles/particles-000* <output_dir> ?

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -131,10 +126,12 @@ namespace aspect
+                                   "The size of the list is the number of spatial dimensions. If only"
+                                   "one value is given, then each spatial dimension is set to the same value."
+                                   "The list of numbers are parsed as a floating point number (so that one can "
                                    "specify, for example, '1e4' particles) but it is interpreted as "
```
I hope nobody tries 1e4 particles per direction in a model with 1e6 cells in 3D :smile:. That would be 1e18 particles. But keep it as an example.

gassmoeller commented on source/particle/generator/reference_cell.cc:
```diff
@@ -160,8 +157,9 @@ namespace aspect
               {
-                for (int i=0; i<dim; i++)
-                  number_of_particles[i] = (unsigned int) (prm.get_double ("Number of tracers per cell per direction"));
+                number_of_particles = Utilities::possibly_extend_from_1_to_N (Utilities::string_to_double(Utilities::split_string_list(prm.get("Number of tracers per cell per direction"))),
```
Here you can introduce a temporary `std::vector<double>` to store the intermediate result and then copy and cast it to `unsigned int` to avoid the casts during the actual particle generation. This could also make the lines somewhat shorter.

hlokavarapu commented:
I have tried renaming the reference files but I ran into a similar issue. I am really not sure what is happening here but nevertheless, I will further investigate this problem and will let you guys know as soon I have an update.

hlokavarapu commented:
Tester is happy. Ready to be reviewed/merged.

gassmoeller commented:
Great, thanks for revising! They are in great shape.
Ready to merge, but feel free to add an entry to changes.h to get some credit for these generators (I will merge tomorrow evening, if you choose not to :smile:).

bangerth commented:
I think there ought to be an entry in the change log, not only for Harsha to get credit, but also for users to see what is new.

tjhei commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -0,0 +1,108 @@
+    {
+      ASPECT_REGISTER_PARTICLE_GENERATOR(QuadraturePoints,
+                                         "quadrature points",
+                                         "Generates particles at the quadrature points of each active cell of"
```
missing space at the end (also at the next line)

tjhei commented on source/particle/generator/quadrature_points.cc:
```diff
@@ -0,0 +1,108 @@
+      ASPECT_REGISTER_PARTICLE_GENERATOR(QuadraturePoints,
+                                         "quadrature points",
+                                         "Generates particles at the quadrature points of each active cell of"
+                                         "the triangulation. Here, Gauss quadrature of degree, (velocity_degree + 1),"
```
remove comma after degree

tjhei replied:
can you please escape ``_`` like this: ``\\_``

tjhei commented on source/particle/generator/reference_cell.cc:
```diff
@@ -0,0 +1,193 @@
+      ASPECT_REGISTER_PARTICLE_GENERATOR(ReferenceCell,
+                                         "reference cell",
+                                         "Generate a uniform distribution of particles per cell and spatial direction in"
+                                         "the unit cell and transforms each of the particles back to real region in the model"
```
again: spaces at the end

gassmoeller commented:
Very nice. Thanks!
